# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/clients/manager/ClientManagerService.cpp
# Commit: 6125bb02952a
# Full Hash: 6125bb02952aff71c0d757e238a7752101b2d6c2
# Author: Jean-Yves Avenard <jyavenard@mozilla.com>
# Date: 2019-09-03 04:06:20
# Regressor Bug: 1556489
# File Overlap Count: 1
# Description:
#   Bug 1556489 - P23 - Only create new ClientSource objects in the content, but use the recreated info from redirects in the parent. r=asuth
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D42260
# ==============================================================================

diff -r 06529d12b589 -r 6125bb02952a dom/clients/manager/ClientManagerService.cpp
--- a/dom/clients/manager/ClientManagerService.cpp	Mon Aug 26 12:18:57 2019 +1000
+++ b/dom/clients/manager/ClientManagerService.cpp	Tue Aug 27 08:53:37 2019 +1000
@@ -199,6 +199,16 @@
     return false;
   }
   entry.OrInsert([&] { return aSource; });
+
+  // Now that we've been created, notify any handles that were
+  // waiting on us.
+  auto* handles = mPendingHandles.GetValue(aSource->Info().Id());
+  if (handles) {
+    for (auto handle : *handles) {
+      handle->FoundSource(aSource);
+    }
+  }
+  mPendingHandles.Remove(aSource->Info().Id());
   return true;
 }
 
@@ -232,6 +242,20 @@
   return source;
 }
 
+void ClientManagerService::WaitForSource(ClientHandleParent* aHandle,
+                                         const nsID& aID) {
+  auto& entry = mPendingHandles.GetOrInsert(aID);
+  entry.AppendElement(aHandle);
+}
+
+void ClientManagerService::StopWaitingForSource(ClientHandleParent* aHandle,
+                                                const nsID& aID) {
+  auto* entry = mPendingHandles.GetValue(aID);
+  if (entry) {
+    entry->RemoveElement(aHandle);
+  }
+}
+
 void ClientManagerService::AddManager(ClientManagerParent* aManager) {
   AssertIsOnBackgroundThread();
   MOZ_DIAGNOSTIC_ASSERT(aManager);